src/File.php

```diff
@@ -485,6 +484,49 @@ public function canDelete($member = null)
             ->canDelete($this->ID, $member);
     }
 
+
+    /**
+     * Check if the File has a CanViewType set or inherited at "LoggedInUsers or above"
+     *
+     * @return bool
+     */
+    public function hasRestrictedAccess(): bool
+    {
+        return $this->hasRestrictedPermissions($this);
+    }
+
+    /**
+     * @param File $file
+     * @return bool
+     */
+    private function hasRestrictedPermissions(File $file): bool
+    {
+        $canViewType = $file->CanViewType;
+        if (in_array($canViewType, [InheritedPermissions::LOGGED_IN_USERS, InheritedPermissions::ONLY_THESE_USERS])) {
+            return true;
+        } elseif ($canViewType == InheritedPermissions::INHERIT) {
+            if ($file->ParentID != 0) {
+                $parent = $file->Parent();
```

Comment:
This will trigger another query for every traversal, right? I'm wondering what the performance is like once you get a few layers deep 🤔

Reply:
That's right, pretty inefficient.  Worse case scenario say subfolder nested 4 deep, 50 paginated records, 200 individual queries.  However I'd expect database caching to kick in as 150 of these queries will be the same as the other 50 queries 'on that level'

Without manually testing performance, I don't know if this is an issue worth solving or not

Options if we need them:
- InheritedPermissions class (I cannot recall my reasoning at the time for choosing not to use this)
- New MyCachingService class that does a few large queries and caches in memory i.e. only for the duration of the request

Reply:
I'm not sure whether `InheritedPermissions` would help here, as it doesn't return specific permission data, just a boolean canView based on a given user.

It'd be great to run some benchmarking against this, as I think it's probably most sensible to leave it as-is, but we'll want to warn users if we expect choppy performance upon initial upgrade.

Reply:
Ugh, performance is pretty bad, even accounting for my slow local

 /admin/graphql request that gets file list

network panel open 'disable cache' (not that that should make a diff)

Root(Files) / 1 folder / 2 folder / 3 folder / 4 folder / 50 Files
(worst case scenario)

xhprof on the 51 requests to hasRestrictedAccess() (50 files + 1 for the current folder)

microsec 1,000,000 = 1 second

with db lookups
648,588
800,023
838,835
752,186
625,229
avg = 732,972

without db lookups, just return false
90

For comparison

Root(Files) / 1 folder / 50 Files
130,664

Reply:
> we expect choppy performance upon initial upgrade

It's straight choppy performance, unlike what we did with the userform lookup where we saved that to the database, we don't save the result of this one because it must stay dynamic

Will investigate `InheritedPermissions` to see if a `Member::actAs(null` will suffice for this (possibly not as we need to test this against draft-content), and failing that will do a custom caching service

Reply:
File::canView() actually already uses InheritedPermissions, though it doesn't quite work as I'd like:

```
# This doesn't work, doesn't read draft site:
# Gets caught up on Versioned::canViewVersioned()
return Versioned::withVersionedMode(function() {
    Versioned::set_reading_mode('Stage.' . Versioned::DRAFT); // same result if Versioned::LIVE
    return Member::actAs(null, function() {
        return !$this->canView();
    });
});

# This works, fast performance, but is it risky?
$default = Versioned::get_draft_site_secured();
Versioned::set_draft_site_secured(false);
$value = Member::actAs(null, function() {
    return !$this->canView();
});
Versioned::set_draft_site_secured($default);
return $value;

2 runs:
11,232
35,774
```

Reply:
Have instead added some basic memory caching into the function.

Works great, 1 run in xhprof was only 4,316